usr/src/uts/armv8/os/ddi_impl.c

```diff
@@ -1157,21 +1157,49 @@ i_ddi_free_intr_phdl(ddi_intr_handle_impl_t *hdlp)
 	hdlp->ih_private = NULL;
 }
 
+
+dev_info_t *
+i_ddi_interrupt_parent(dev_info_t *pdip)
+{
+	int len;
+
+	while (pdip != NULL) {
+		phandle_t ph;
+
+		if (ddi_prop_get_int(DDI_DEV_T_ANY, pdip, 0,
+		    "#interrupt-cells", -1) != -1) {
+			return (pdip);
+		}
+
+		if ((ph = ddi_prop_get_int(DDI_DEV_T_ANY, pdip, 0,
+		    "interrupt-parent", -1)) != -1) {
+			pnode_t node = prom_findnode_by_phandle(ph);
+			pdip = e_ddi_nodeid_to_dip(node);
+			continue;
+		}
+
+		pdip = ddi_get_parent(pdip);
+	}
+
+	return (NULL);
+}
+
 int
 i_ddi_get_intx_nintrs(dev_info_t *dip)
 {
+	dev_info_t *ipar = i_ddi_interrupt_parent(dip);
 	uint_t intrlen;
 	int intr_sz;
 	int *ip;
 	int ret = 0;
 
-	if (ddi_prop_lookup_int_array(DDI_DEV_T_ANY, dip, DDI_PROP_DONTPASS |
-	    DDI_PROP_CANSLEEP,
+	VERIFY3P(ipar, !=, NULL);
+
+	if (ddi_prop_lookup_int_array(DDI_DEV_T_ANY, dip, DDI_PROP_DONTPASS,
 	    "interrupts", &ip, &intrlen) == DDI_SUCCESS) {
-		intr_sz = ddi_prop_get_int(DDI_DEV_T_ANY, dip,
+		intr_sz = ddi_prop_get_int(DDI_DEV_T_ANY, ipar,
```

Comment:
I'm not sure this is quite right. The spec states that `#interrupt-cells` is used at the root of an interrupt domain, which in turn is described as:
_An interrupt domain is the context in which an interrupt specifier is interpreted. The root of the domain is either (1) an interrupt controller or(2) an interrupt nexus._
1. _An interrupt controller is a physical device and will need a driver to handle interrupts routed through it. It may also cascade into another interrupt domain. An interrupt controller is specified by the presence of an interrupt-controller property on that node in the devicetree._
2. _An interrupt nexus defines a translation between one interrupt domain and another. The translation is based on both domain-specific and bus-specific information. This translation between domains is performed with the interrupt-map property. For example, a PCI controller device node could be an interrup tnexus that defines a translation from the PCI interrupt namespace (INTA,INTB, etc.) to an interrupt controller with Interrupt Request (IRQ) numbers._

So it all gets very complicated, since we'd really need to traverse up the interrupt hierarchy until we hit the first controller or nexus (i.e. a node that defines `#interrupt-cells`.